[BPK-2074] Add accessory view to Badge

georgegillams opened the pull request.

willhg commented:
LGTM

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge-test.common.js:
```diff
@@ -19,14 +19,26 @@
+    color: colorWhite,
+  },
+});
+
```
These styles will be unnecessary once we change it to automatically colour the accessory view.

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -38,6 +38,9 @@ import { View, ViewPropTypes, StyleSheet } from 'react-native';
 
 const styles = StyleSheet.create({
   viewBase: {
+    display: 'flex',
```
Isn't `display: 'flex'` the default? You may be able to remove this line and it'll still work - worth a go!

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -97,6 +100,9 @@ const styles = StyleSheet.create({
     borderBottomRightRadius: 0,
   },
+  accessoryView: {
+    marginRight: spacingSm,
```
`marginEnd` is better - works with RTL then.

georgegillams replied:
Strangely it seemed to work correctly with `marginRight`, but I'll use RTL friendly ones just to be sure

<img width="545" alt="screenshot 2018-11-22 at 17 37 29" src="https://user-images.githubusercontent.com/30267516/48917174-4faf2080-ee7d-11e8-9376-c57c3e6cd704.png">

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -157,6 +164,9 @@ const BpkBadge = (props: Props) => {
     <View style={viewStyle}>
+      {accessoryView && (
+        <View style={styles.accessoryView}>{accessoryView}</View>
+      )}
```
This wrapper view is only being used to add the margin style. Could we avoid an extra view by conditionally applying the margin to the text instead, when the accessory view is supplied?

shaundon replied:
I spoke to @willhg and we agreed that users shouldn't have to (or indeed be able to) colour the accessory view themselves, so the badge component should do this for them. 

We can do this by cloning the children and colouring any `Text` or `BpkIcon` components to match the badge's text colour.

Here's some examples of where we've iterated over or cloned children before that you could base it off:

1. [Cloning images passed to flat list items to add a margin](https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-flat-list/src/BpkFlatListItem.android.js#L74)
1. [Iterating over children passed into pickers to grab some data](https://github.com/Skyscanner/backpack/blob/master/native/packages/react-native-bpk-component-picker/src/BpkPickerMenu.android.js#L76)

shaundon commented on native/packages/react-native-bpk-component-badge/stories.js:
```diff
@@ -64,13 +67,22 @@ const style = StyleSheet.create({
+  },
+  iconOutline: {
+    color: colorWhite,
+  },
```
These styles will be unnecessary once we change it to automatically colour the accessory view.

shaundon commented on packages/bpk-docs/src/pages/NativeBadgePage/NativeBadgePage.js:
```diff
@@ -51,6 +53,28 @@ const components = [
     ],
   },
+  {
+    id: 'withAccessoryView',
```
Convention for IDs is kebab case: `with-accessory-view`.

shaundon replied:
The link in `unreleased.md` will need to change accordingly too.

georgegillams replied:
Great catch for `UNRELEASED.md`

shaundon commented:
Oh just noticed, `lodash` isn't in the dependencies of this package. Although I think you'll end up removing the lodash-using parts anyway, so.. 🤷‍♂️

georgegillams commented:
I've addressed your feedback now @shaundon 

<img width="300" alt="screenshot 2018-11-23 at 07 12 21" src="https://user-images.githubusercontent.com/30267516/48931820-2d0a1f80-eef0-11e8-99c1-68b169f70f9b.png">
<img width="300" alt="screenshot 2018-11-23 at 07 12 04" src="https://user-images.githubusercontent.com/30267516/48931821-2d0a1f80-eef0-11e8-8cfd-a0d14380afa4.png">

Within the badge component itself, I've applied styling to the `accessoryView` as suggested.
This works only for 1 `accessoryView` component, however - and I'm not sure how we should go about adjusting multiple `accessoryView` components.
 - I thought about recursively iterating through the `accessoryView`'s children and apply the styling to each descendant - but this seems like quite hacky behaviour, and will probably be surprising to consumers.
 - Alternatively I could change the API to accept an `accessoryViews` array, and apply the styling to each element in there

Let me know what you think is best

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -97,6 +100,15 @@ const styles = StyleSheet.create({
+  },
+  accessoryViewOutline: {
+    color: colorWhite,
+  },
```
Can't we reuse `textBase`, `textDestructive` and `textOutline` styles for the accessory view?

georgegillams replied:
Yo - I tried this but Flow starts complaining at me 😢

shaundon commented on UNRELEASED.md:
```diff
@@ -1,5 +1,10 @@
+
+ - react-native-bpk-component-badge:
+     - Badges now accept an optional `accessoryView` that is displayed on the leading side, and can contain any content including icons. See https://backpack.github.io/components/badge/?platform=native#with-multiple-icons
+
```
This'll need to change as the props have changed too.

shaundon commented on native/packages/react-native-bpk-component-badge/README.md:
```diff
@@ -44,8 +45,10 @@ export default class App extends Component {
 | --------------------- | --------------------------------------------------------------------------| -------- | ------------- |
 | message               | string                                                                    | true     | -             |
 | docked                | oneOf('start', 'end')                                                     | false    | null          |
+| iconSeparator         | string                                                                    | false    | null          |
```
Let's make sure this is well documented. I was thinking we could have a heading underneath this table, something like:

```
## iconSeparator

`iconSeparator` is an optional string to place between icons when there are more than one, for example when you want to place a "+" between the icons.
```

shaundon commented on native/packages/react-native-bpk-component-badge/README.md:
```diff
@@ -44,8 +45,10 @@ export default class App extends Component {
 | message               | string                                                                    | true     | -             |
 | docked                | oneOf('start', 'end')                                                     | false    | null          |
+| iconSeparator         | string                                                                    | false    | null          |
+| icons                 | arrayOf(string)                                                           | false    | []            |
```
Can we make this more clear that it needs to be an icon string specifically? `arrayOf(BpkIcon.icons)` or something?

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -166,15 +194,19 @@ const BpkBadge = (props: Props) => {
 
 BpkBadge.defaultProps = {
-  type: BADGE_TYPES.warning,
+  icons: [],
```
I think the default should be `null` - no need to use an array if it isn't needed (remember to update prop types too if you do this).

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -157,6 +169,22 @@ const BpkBadge = (props: Props) => {
   return (
     <View style={viewStyle}>
+      {icons &&
+        icons.map((i, n) => (
```
One-letter variable names can be a bit obtuse - can we use `iconName` and `index` instead?

georgegillams replied:
I agree - I hate it when people write code like this. Goddamnit!

shaundon commented on native/packages/react-native-bpk-component-badge/stories.js:
```diff
@@ -21,11 +21,13 @@
   colorGray700,
   spacingSm,
 } from 'bpk-tokens/tokens/base.react.native';
+import { capitalize } from 'lodash';
```
Why are we adding an import instead of doing it in a one-liner that we already had?

georgegillams replied:
Why write a one-liner that we don't already have when we could just import it from a library we do already have 😆

shaundon replied:
Counter-counter question/statement: You **don't** already have Lodash; it isn't in this package's dependencies 😉

georgegillams replied:
Counter-counter-counter statement: Yeah but this is stories 🤷‍♂️ 
So we do have Lodash

shaundon replied:
Counter-counter-counter-counter statement: It's only working because you have Lodash installed elsewhere - by that logic why do we add anything at all to `devDependencies` 😛

georgegillams replied:
So what d'yall suggest then?

shaundon commented on packages/bpk-docs/src/pages/NativeBadgePage/NativeBadgePage.js:
```diff
@@ -51,6 +55,50 @@ const components = [
+  },
+  {
+    id: 'with-icons',
+    title: 'With multiple icons',
```
The icon separator is optional, so let's make it clear it's been supplied here.
```suggestion
    title: 'With multiple icons and an icon separator',
```

georgegillams commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -157,6 +169,26 @@ const BpkBadge = (props: Props) => {
+            <BpkIcon
+              small
+              icon={icon}
+              style={JSON.parse(JSON.stringify(textStyle))}
```
For some reason, this fixes the flow issue I was seeing.
Don't know why - this isn't really a solution!

georgegillams commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -155,8 +159,13 @@ const BpkBadge = (props: Props) => {
     viewStyle.push(userStyle);
   }
 
+  const adjustAccessoryView = React.cloneElement(accessoryView, {
```
Will break if `accessoryView === null`

georgegillams commented on native/packages/react-native-bpk-component-badge/src/BpkBadge.js:
```diff
@@ -115,6 +118,7 @@ export const BADGE_DOCKED_TYPES = {
 
 export type Props = {
   message: string,
+  accessoryView: ?Element<BpkBadgeIcons>,
```
Should this be less limited?

shaundon commented on UNRELEASED.md:
```diff
@@ -1 +1,6 @@
+
+**Added:**
+ - react-native-bpk-component-badge:
+     - Badges now accept an optional `accessoryView` that allows icons, or other content, to be placed alongside the text inside the bade. See https://backpack.github.io/components/badge/?platform=native#with-multiple-icons
```
You changed the ID in the docs to `with-icons` so the link will need to change too.

georgegillams replied:
🤦‍♂️

shaundon commented on native/packages/react-native-bpk-component-badge/README.md:
```diff
@@ -34,18 +35,36 @@ export default class App extends Component {
 | type                  | oneOf('success', 'warning', 'destructive', 'light', 'inverse', 'outline') | false    | warning       |
+
+
+## Accessory view
```
```suggestion
## `accessoryView`
 The accessory view allows for icons to be placed in front of the text inside the Badge component in conjunction with `BpkBadgeIcons`.

## `BpkBadgeIcons`

Renders icons suitable to be placed inside a Badge.

### Props

<props table here for BpkBadgeIcons here>
```

I based this off of https://backpack.github.io/components/navigation-bar?platform=native

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadgeIcons.js:
```diff
@@ -0,0 +1,123 @@
+  },
+});
+
+export const BADGE_TYPES = {
```
Unused - remove.

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadgeIcons.js:
```diff
@@ -0,0 +1,123 @@
+  outline: 'outline',
+};
+
+export const BADGE_DOCKED_TYPES = {
```
Unused - remove.

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadgeIcons.js:
```diff
@@ -0,0 +1,123 @@
+    justifyContent: 'center',
+  },
+  item: {
+    color: colorGray700,
```
Default colour not required.

shaundon commented on UNRELEASED.md:
```diff
@@ -1 +1,6 @@
+
+**Added:**
+ - react-native-bpk-component-badge:
+     - Badges now accept an optional `accessoryView` that allows icons, or other content, to be placed alongside the text inside the bade. See https://backpack.github.io/components/badge/?platform=native#with-icons
```
```suggestion
     - Badges now accept an optional `accessoryView` that allows icons, or other content, to be placed alongside the text inside the badge. See https://backpack.github.io/components/badge/?platform=native#with-icons
```

shaundon commented on native/packages/react-native-bpk-component-badge/README.md:
```diff
@@ -34,18 +35,36 @@ export default class App extends Component {
+| Property              | PropType                                                                  | Required | Default Value |
 | --------------------- | --------------------------------------------------------------------------| -------- | ------------- |
 | message               | string                                                                    | true     | -             |
+| accessoryView         | node                                                                      | false    | null          |
```
Should be `element`.

shaundon commented on native/packages/react-native-bpk-component-badge/src/BpkBadgeIcons.js:
```diff
@@ -0,0 +1,92 @@
+  icons: Array<Element<typeof BpkIcon>>,
+  itemStyle: ?(Object | Array<Object>),
+  separator: ?string,
+  style: ?(Object | Array<Object>),
```
`style` should be removed - it's not applied anywhere.

shaundon commented on native/packages/react-native-bpk-component-badge/README.md:
```diff
@@ -34,18 +35,36 @@ export default class App extends Component {
 | type                  | oneOf('success', 'warning', 'destructive', 'light', 'inverse', 'outline') | false    | warning       |
+
+
+## `accessoryView`
```
https://github.com/Skyscanner/backpack/pull/1173#discussion_r236237279

shaundon commented on native/packages/react-native-bpk-component-badge/README.md:
```diff
@@ -34,18 +35,46 @@ export default class App extends Component {
+
+| Property  | PropType                | Required | Default Value |
+| --------- | ------------------------| -------- | ------------- |
+| icons     | arrayOf(oneOf(BpkIcon)) | true     | -             |
```
```suggestion
| icons     | arrayOf(BpkIcon) | true     | -             |
```
